File: game/src/behaviours/blocks.cpp
#include "behaviours/blocks.h"

#include "behaviours/rng.h"

#include <algorithm>
#include <cmath>

#include "components.h"  // kInventoryCap

namespace badlands {

namespace {

// Scores are CONSIDERATIONS in [0,1] -- "how much does the situation call for
// this right now" -- never priorities and never preferences. There is no
// worthiness ranking of activities anywhere: within the Normal band a hero
// rests rather than hunts because it is TIRED, and stops doing so the moment it
// is not.
//
// For a need-driven activity the score is the need's URGENCY (see urgency()).
// For one that is simply available or not, it is kApplies / kNotApplicable, and
// the per-class weight scales it into the same contest.
constexpr float kApplies = 1.0f;
constexpr float kNotApplicable = 0.0f;

// How badly a depleted reserve wants attention: 0 at or above `threshold`,
// ramping linearly to 1 when the reserve is empty. One shape, used by every
// need, with the threshold as the per-need dial.
float urgency(float reserve, float threshold) {
    if (threshold <= 0.0f || reserve >= threshold) {
        return kNotApplicable;
    }
    return std::clamp((threshold - reserve) / threshold, 0.0f, 1.0f);
}

// Townfolk still run select_priority (first applicable in list order), so only
// the sign of their scores matters, not the magnitude.
constexpr float kTierVisitTax = kApplies;
constexpr float kTierDeposit = kApplies;

}  // namespace

// --- GoHome (rest) ----------------------------------------------------------
// Scored on three things, exactly as specified: how spent the hero is, whether
// it is the sleep window, and whether it is hurt. The night raises the bar at
// which turning in starts to appeal (a hero will go to bed merely tired-ish
// after dark, but pushes on through the day); injury urges rest on its own,
// whatever the reserve says -- hence max() rather than a product, which would
// let a well-rested hero bleed out on its feet.
float score_go_home(const WorldView& v, const SimFactors& f) {
    if (!v.has_home) {
        return kNotApplicable;
    }
    const float bar = v.night ? f.hero.fatigue_seek_night : f.hero.fatigue_seek;
    return std::max(urgency(v.fatigue, bar), urgency(v.health_frac, f.hero.low_health_rest));
}
BehaviourResult act_go_home(const WorldView& v, const SimFactors&) {
    return {Behavior::GoHome, v.home_door, Command{CommandKind::EnterHome, v.slot}, true};
}

// --- Buy --------------------------------------------------------------------
float score_buy(const WorldView& v, const SimFactors&) {
    return (v.has_apothecary && v.inventory < kInventoryCap) ? kApplies : kNotApplicable;
}
BehaviourResult act_buy(const WorldView& v, const SimFactors&) {
    return {Behavior::Buy, v.apothecary_door, Command{CommandKind::Buy, v.slot}, true};
}

// --- VisitTavern ------------------------------------------------------------
float score_visit_tavern(const WorldView& v, const SimFactors& f) {
    if (!v.has_tavern || v.night) {
        return kNotApplicable;  // shut after dark
    }
    return urgency(v.content, f.hero.content_seek);
}
BehaviourResult act_visit_tavern(const WorldView& v, const SimFactors&) {
    Command enter{CommandKind::EnterBuilding, v.slot, UINT32_MAX, {0.0f, 0.0f},
                  static_cast<int32_t>(BuildingKind::Tavern)};
    return {Behavior::VisitTavern, v.tavern_door, enter, true};
}

// --- Chat -------------------------------------------------------------------
float score_chat(const WorldView& v, const SimFactors& f) {
    if (v.chatting) {
        return kApplies;  // mid-conversation: see it through
    }
    if (!v.has_chat_partner) {
        return kNotApplicable;
    }
    return urgency(v.content, f.hero.chat_content_seek);
}
BehaviourResult act_chat(const WorldView& v, const SimFactors& f) {
    if (v.chatting) {
        return {Behavior::Chat, v.pos, std::nullopt, false};  // stand and talk
    }
    // Walk over, and strike it up once close enough. The handler re-validates
    // (both present, in range, neither already engaged), so emitting on arrival
    // keeps the approach itself out of the command log.
    BehaviourResult r{Behavior::Chat, v.partner_pos, std::nullopt, false};
    if (v.partner_dist <= f.hero.chat_radius) {
        r.follow_up = Command{CommandKind::Chat, v.slot, v.partner_slot};
    }
    return r;
}

// --- Explore ----------------------------------------------------------------
float score_explore(const WorldView& v, const SimFactors& f) {
    if (!v.has_explore_goal) {
        return kNotApplicable;  // nowhere unknown within reach, or not in the mood
    }
    if (v.move_blocked) {
        return kNotApplicable;  // the world said no; try elsewhere next window
    }
    if (v.fatigue <= f.hero.explore_min_fatigue) {
        return kNotApplicable;  // not enough in the tank -- stay near home
    }
    if (v.has_prey) {
        return kNotApplicable;  // something worth stopping for is right here
    }
    return kApplies;
}
BehaviourResult act_explore(const WorldView& v, const SimFactors&) {
    return {Behavior::Explore, v.explore_goal, std::nullopt, false};
}

// --- Hunt (hunter) ----------------------------------------------------------
float score_hunt(const WorldView& v, const SimFactors&) {
    return v.has_prey ? kApplies : kNotApplicable;
}
BehaviourResult act_hunt(const WorldView& v, const SimFactors&) {
    BehaviourResult r{Behavior::Hunt, v.prey_pos, std::nullopt, false};
    // Chase to prey_pos; once within the hunter's own reach, take the shot (the
    // handler re-checks range + cooldown, so the log gets one entry per shot).
    if (v.prey_dist <= v.self_attack_range) {
        r.follow_up = Command{CommandKind::Attack, v.slot, v.prey_slot};
    }
    return r;
}

// --- Roam (shared) ----------------------------------------------------------
glm::vec2 roam_point(uint32_t slot, int64_t epoch, glm::vec2 anchor, float radius) {
    // Seeded off the slot and the roam epoch so the goal is stable within a
    // lease window yet unique per entity -- deterministic, no wall-clock, no
    // global RNG state. Shares seed_of with every other draw in the sim, so the
    // same avalanche that fixed the exploration appetite applies here (this
    // used to keep its own weaker copy of the mix).
    uint64_t s = seed_of(slot, epoch);
    const float ang = unit_float(s) * 6.2831853f;
    const float rad = unit_float(s) * radius;
    return anchor + glm::vec2{std::cos(ang) * rad, std::sin(ang) * rad};
}
float score_roam(const WorldView&, const SimFactors&) { return kApplies; }
BehaviourResult act_roam(const WorldView& v, const SimFactors&) {
    return {Behavior::Roam, v.roam_goal, std::nullopt, false};
}

// --- Flee (shared) ----------------------------------------------------------
float score_flee(const WorldView& v, const SimFactors& f) {
    return (has_threat(v) && nearest_threat_dist(v) <= f.critter.flee_radius) ? kApplies
                                                                              : kNotApplicable;
}
BehaviourResult act_flee(const WorldView& v, const SimFactors& f) {
    glm::vec2 away = v.pos - nearest_threat_pos(v);
    const float len = glm::length(away);
    away = (len > 1e-4f) ? away / len : glm::vec2{1.0f, 0.0f};  // degenerate: pick a dir
    // Reports its own id rather than masquerading as Roam: a bolt and a wander
    // are different goals, and the statistics histogram must be able to tell
    // "the herd is panicking" from "the herd is grazing".
    return {Behavior::Flee, v.pos + away * f.critter.flee_distance, std::nullopt, false};
}

// --- Idle -------------------------------------------------------------------
float score_idle(const WorldView&, const SimFactors&) { return kApplies; }
BehaviourResult act_idle(const WorldView& v, const SimFactors&) {
    return {Behavior::Idle, v.pos, std::nullopt, false};
}

// --- Graze (critter) --------------------------------------------------------
float score_graze(const WorldView& v, const SimFactors&) {
    return v.grazing ? kApplies : kNotApplicable;
}
BehaviourResult act_graze(const WorldView& v, const SimFactors&) {
    return {Behavior::Graze, v.pos, std::nullopt, false};  // hold and feed
}

// --- VisitNextTaxable (townfolk) --------------------------------------------
float score_visit_taxable(const WorldView& v, const SimFactors&) {
    return v.has_tax_target ? kTierVisitTax : kNotApplicable;
}
BehaviourResult act_visit_taxable(const WorldView& v, const SimFactors&) {
    Command collect{CommandKind::CollectTax, v.slot, v.tax_target_id};
    return {Behavior::VisitTax, v.tax_target_door, collect, true};
}

// --- Deposit (townfolk) -----------------------------------------------------
float score_deposit(const WorldView& v, const SimFactors&) {
    return v.has_deposit ? kTierDeposit : kNotApplicable;
}
BehaviourResult act_deposit(const WorldView& v, const SimFactors&) {
    return {Behavior::Deposit, v.deposit_door, Command{CommandKind::Deposit, v.slot}, true};
}

}  // namespace badlands

File: scripts/brains/nim/blocks.nim
# Port of the hero score_*/act_* pairs (game/src/behaviours/blocks.cpp)
# actually reachable through kHeroActivities (game/src/town_brain.cpp), now
# emitting a v2 Suggestion (BL_INT_*) instead of v1's Decision
# (activityId/goalKind/BL_CMD_*).
#
# A `score_*` returns a CONSIDERATION in [0,1] (0 = veto) -- UNCHANGED from
# v1, byte-for-byte the same comparisons/thresholds. `act_*` differs from v1
# in shape, not spirit: v1's engine (apply_brain_decision) gated a
# "walk-then-fire-follow-up-on-arrival" itself, generically, for every
# activity; the v2 engine (apply_intention, game/src/intention.h) has no such
# combined walk+gate for ENTER/ENTER_HOME/BUY -- each intention kind is
# either a walk (MoveTo) or a bare action, never both. So the three
# door-activities (GoHome/Buy/VisitTavern) now do that arrival check
# THEMSELVES, against BlViewFactors.entrance_radius (mirrors game/src/
# components.h's kEntranceRadius): far -> suggest MoveTo toward the door;
# close enough -> suggest the actual action. Chat/Hunt already carried their
# own precise gate in v1 (chat_radius/self_attack_range, both still on the
# wire) and are unchanged in spirit.

import abi
import activity_catalog
import hero_view

# Scores are considerations, never priorities/preferences (see blocks.cpp's
# top-of-file comment) -- kApplies/kNotApplicable name the two flat cases.
const
  kApplies: float32 = 1.0'f32
  kNotApplicable: float32 = 0.0'f32

  # components.h: constexpr int kInventoryCap = 2 (elixirs a hero can carry).
  kInventoryCap: int32 = 2

  # badlands_sim.hpp's BuildingKind: Castle=0, FreeCompanyQuarters=1,
  # HuntersCamp=2, ThievesDen=3, Scriptorium=4, Tavern=5.
  kBuildingKindTavern: int32 = 5

# How close to a door an act_* block below must be before firing the actual
# action (Enter/EnterHome/Buy) instead of just walking toward it -- read from
# the wire's own BlViewFactors.entrance_radius (mirrors game/src/components.h's
# kEntranceRadius; was a hand-copied constant here, now the wire is the single
# source of truth -- see brain_abi.h's BlViewFactors comment). v1 had this
# check done GENERICALLY, host-side, by apply_brain_decision's
# follow_up_on_arrival gate; v2's apply_intention has no such combined gate
# (see this file's top comment), so it moves here. Compared as a squared
# distance to avoid pulling in sqrt/libm at all -- one fewer thing for
# scripts/build_brains.sh's WASI-import bisect to worry about.
proc entranceRadiusSq(f: BlViewFactors): float32 =
  f.entrance_radius * f.entrance_radius

# How badly a depleted reserve wants attention: 0 at/above `threshold`, ramping
# linearly to 1 when empty. One shape, used by every need (blocks.cpp: urgency).
proc urgency(reserve, threshold: float32): float32 =
  if threshold <= 0.0'f32 or reserve >= threshold:
    return kNotApplicable
  let x = (threshold - reserve) / threshold
  result = if x < 0.0'f32: 0.0'f32 elif x > 1.0'f32: 1.0'f32 else: x

proc distSq(a, b: Vec2): float32 =
  let dx = a.x - b.x
  let dz = a.z - b.z
  result = dx * dx + dz * dz

# One activity's suggestion: BL_INT_* + the fields it uses (mirrors
# BlSuggestionWire's own shape, minus duration_millis/idle_hint_millis,
# which hero.nim's top-level react() fills in once, not per-activity).
type Suggestion* = object
  kind*: int32              # BL_INT_*
  activityLabel*: int32     # ActivityId, inspection only
  pointX*, pointZ*: float32
  targetSlot*: uint32
  arg*: int32

proc moveTo(activityLabel: int32, goal: Vec2): Suggestion =
  Suggestion(kind: BL_INT_MOVE_TO, activityLabel: activityLabel, pointX: goal.x, pointZ: goal.z)

# --- GoHome (rest) -----------------------------------------------------------
proc scoreGoHome*(v: HeroView, f: BlViewFactors): float32 =
  if not v.hasHome:
    return kNotApplicable
  let bar = if v.night: f.fatigue_seek_night else: f.fatigue_seek
  result = max(urgency(v.fatigue, bar), urgency(v.healthFrac, f.low_health_rest))

proc actGoHome*(v: HeroView, f: BlViewFactors): Suggestion =
  if distSq(v.pos, v.homeDoor) <= entranceRadiusSq(f):
    return Suggestion(kind: BL_INT_ENTER_HOME, activityLabel: ActGoHome)
  result = moveTo(ActGoHome, v.homeDoor)

# --- Buy ----------------------------------------------------------------------
proc scoreBuy*(v: HeroView, f: BlViewFactors): float32 =
  result = if v.hasApothecary and v.inventory < kInventoryCap: kApplies else: kNotApplicable

proc actBuy*(v: HeroView, f: BlViewFactors): Suggestion =
  if distSq(v.pos, v.apothecaryDoor) <= entranceRadiusSq(f):
    return Suggestion(kind: BL_INT_BUY, activityLabel: ActBuy)
  result = moveTo(ActBuy, v.apothecaryDoor)

# --- VisitTavern ---------------------------------------------------------------
proc scoreVisitTavern*(v: HeroView, f: BlViewFactors): float32 =
  if not v.hasTavern or v.night:
    return kNotApplicable  # shut after dark
  result = urgency(v.content, f.content_seek)

proc actVisitTavern*(v: HeroView, f: BlViewFactors): Suggestion =
  if distSq(v.pos, v.tavernDoor) <= entranceRadiusSq(f):
    return Suggestion(kind: BL_INT_ENTER, activityLabel: ActVisitTavern, arg: kBuildingKindTavern)
  result = moveTo(ActVisitTavern, v.tavernDoor)

# --- Chat -----------------------------------------------------------------------
proc scoreChat*(v: HeroView, f: BlViewFactors): float32 =
  if v.chatting:
    return kApplies  # mid-conversation: see it through
  if not v.hasChatPartner:
    return kNotApplicable
  result = urgency(v.content, f.chat_content_seek)

proc actChat*(v: HeroView, f: BlViewFactors): Suggestion =
  if v.chatting:
    # Nothing NEW to decide: the conversation runs to its own clock
    # (ChattingState + advance_chats, engine-side), and observe_hero never
    # populates partner_slot while already chatting (v.chatting excludes the
    # nearest_companion scan), so there is no valid target to re-suggest Chat
    # with even if we wanted to. BL_INT_IDLE here (not BL_INT_NONE) is
    # deliberate, not a mistake: apply_intention's BL_INT_NONE path leaves
    # CurrentIntention untouched entirely, including its wake_at_millis --
    # which would leave a chatting hero's wake schedule stuck at whatever it
    # was BEFORE the conversation started, so once that deadline passed it
    # would re-wake the brain every single tick for as long as the chat
    # continues. Idle's own duration_millis (the idle hint) keeps a fresh,
    # bounded re-check window instead -- activityLabel still reports Chat for
    # inspection, so the histogram/HeroSimulationState.behavior are unaffected.
    return Suggestion(kind: BL_INT_IDLE, activityLabel: ActChat)
  # Walk over, and strike it up once close enough -- the engine still gates
  # on chat_radius itself (command.cpp's Chat handler), this just avoids
  # suggesting Chat before it could possibly be valid.
  if v.partnerDist <= f.chat_radius:
    return Suggestion(kind: BL_INT_CHAT, activityLabel: ActChat, targetSlot: v.partnerSlot)
  result = moveTo(ActChat, v.partnerPos)

# --- Explore ---------------------------------------------------------------------
proc scoreExplore*(v: HeroView, f: BlViewFactors): float32 =
  if not v.hasExploreGoal:
    return kNotApplicable  # nowhere unknown within reach, or not in the mood
  if v.moveBlocked:
    return kNotApplicable  # the world said no; try elsewhere next window
  if v.fatigue <= f.explore_min_fatigue:
    return kNotApplicable  # not enough in the tank -- stay near home
  if v.hasPrey:
    return kNotApplicable  # something worth stopping for is right here
  result = kApplies

proc actExplore*(v: HeroView, f: BlViewFactors): Suggestion =
  result = moveTo(ActExplore, v.exploreGoal)

# --- Hunt (hunter) -----------------------------------------------------------------
proc scoreHunt*(v: HeroView, f: BlViewFactors): float32 =
  result = if v.hasPrey: kApplies else: kNotApplicable

proc actHunt*(v: HeroView, f: BlViewFactors): Suggestion =
  # Chase to prey_pos; once within the hunter's own reach, take the shot (the
  # handler re-checks range + cooldown, so the log gets one entry per shot).
  if v.preyDist <= v.selfAttackRange:
    return Suggestion(kind: BL_INT_SHOOT, activityLabel: ActHunt, targetSlot: v.preySlot)
  result = moveTo(ActHunt, v.preyPos)

# --- Roam (shared) -------------------------------------------------------------------
# Unlike blocks.cpp, this does NOT re-derive the wander point: observe_hero
# already drew it host-side (roam_point) and shipped it as suggest.roam_goal_*.
proc scoreRoam*(v: HeroView, f: BlViewFactors): float32 = kApplies

proc actRoam*(v: HeroView, f: BlViewFactors): Suggestion =
  result = moveTo(ActRoam, v.roamGoal)

# --- Idle -----------------------------------------------------------------------------
proc scoreIdle*(v: HeroView, f: BlViewFactors): float32 = kApplies

proc actIdle*(v: HeroView, f: BlViewFactors): Suggestion =
  # duration_millis is filled in by hero.nim's top-level react() (the idle
  # hint IS the duration now -- see its own comment).
  Suggestion(kind: BL_INT_IDLE, activityLabel: ActIdle)
